Краткое инфо по проекту для меня-будущего:<br>
* Есть быстромутирующий вирус, мы можем найти редкие мутации (которые появляются, но всего лишь в <1% ридов после секвенирования). Как отличить мутацию от ошибки секвенирования? <br>
* Берём контрольные образцы (100% чистый кусок вируса, встроенный в плазмиду и несколько раз секвенированный) $-$ мы стопроц уверены в их последовательности, а значит все "редкие мутации", которые мы найдём в контролях $-$ это ошибки метода. <br>
* Сравниваем те редкие мутации, которые мы нашли в real-world данных с редкими мутациями из контрольных образцов. Отсеиваем лишние. Как сравнивать? Например по частоте (напр., в контрольных образцах все редкие мутации будут с частотой 0.25%, а в real-world данных чуть чаще (0.8%))

$~$

$~$
P.S. Если вы нашли какой-то интересный метод отличать ошибки от редких мутаций -- напишите его пожалуйста в рецензии, я буду очень благодарен! <з

$~$

## 1. Inspect the data from your roommate

In [1]:
# download roommate's reads
!wget -c http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/SRR1705851.fastq.gz

--2021-10-24 16:44:46--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/SRR1705851.fastq.gz
Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.197.74
Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.197.74|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 19217176 (18M) [application/octet-stream]
Saving to: ‘SRR1705851.fastq.gz’

SRR1705851.fastq.gz   5%[>                   ]   1,00M   333KB/s    in 56s     

2021-10-24 16:45:43 (18,2 KB/s) - Connection closed at byte 1048576. Retrying.

--2021-10-24 16:45:44--  (try: 2)  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/001/SRR1705851/SRR1705851.fastq.gz
Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.197.74|:80... connected.
HTTP request sent, awaiting response... 206 Partial Content
Length: 19217176 (18M), 18168600 (17M) remaining [application/octet-stream]
Saving to: ‘SRR1705851.fastq.gz’

SRR1705851.fastq.gz 100%[+==================>]  18,33M   748KB/s    in 88s     

2021-10-

In [2]:
# rename to roommate.fastq and move to raw_data/
!mv SRR1705851.fastq.gz raw_data/roommate.fastq.gz

In [3]:
# unpack (-k for saving .gz file)
!gunzip -k raw_data/roommate.fastq.gz

In [19]:
# visual check
!head -10 raw_data/roommate.fastq

@SRR1705851.1 1/1
TTCGTGATTGTTTTCACTATCGTTCCGTTTGGCACTGCATGGTGCCCAAGGCACAGCGTTGCCGTGCTGTTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGTTAATAG
+
,<==<<<<A@@@@@@@EEE;CEE+AC>EC;>EFFDC@=A@AE999DDD>>@E777EE75C>EF>EDEEFFFF--AE>EDEEEED=C-58AE=<D=<<DD=D9CDD@EEDED@DEDDE*9;@DDED@@@7@E*;*888@*8;@8@;;@@E
@SRR1705851.2 2/1
NATTAACCATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCGCTCAAAAACTTCCTGGAAATGACAACAGCACGGCAACGCTGTGCCTTGGGCACCATGCAGTGCCAAACGGAACGATAGTGAAAACAATCACGAATGA
+
#5<???BBEEEDEDDDGGGGGGIIIIIIIIIIIIIIIIIIIIIHIIIIFHHIIHHHHHIIIIHIIIIIIIHIIIIIIIIIIIIIIHHHHHHHHHHEHHHHHFFHHHHHHFFHHGFGGGGGGGGGGGGGEEEGCEEGGGGGEEGGGGCGEGG
@SRR1705851.3 3/1
GTTTGGCACTGCATGGTGCCCAAGGCACAGCGTTGCCGTGCTGTTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGTTAATAG


In [20]:
#
!mkdir 1_fastqc
!cp raw_data/roommate.fastq 1_fastqc/
# ...check reads' quality
!fastqc 1_fastqc/roommate.fastq

Started analysis of roommate.fastq
Approx 5% complete for roommate.fastq
Approx 10% complete for roommate.fastq
Approx 15% complete for roommate.fastq
Approx 20% complete for roommate.fastq
Approx 25% complete for roommate.fastq
Approx 30% complete for roommate.fastq
Approx 35% complete for roommate.fastq
Approx 40% complete for roommate.fastq
Approx 45% complete for roommate.fastq
Approx 50% complete for roommate.fastq
Approx 55% complete for roommate.fastq
Approx 60% complete for roommate.fastq
Approx 65% complete for roommate.fastq
Approx 70% complete for roommate.fastq
Approx 75% complete for roommate.fastq
Approx 80% complete for roommate.fastq
Approx 85% complete for roommate.fastq
Approx 90% complete for roommate.fastq
Approx 95% complete for roommate.fastq
Analysis complete for roommate.fastq


#### Results of checking reads quality:
* Quality is good;
* Read len ~35-151 b.p.;
* Total 358265 reads;
* GC-content 42%

In [23]:
# file structure
!ls --width=1 -R --color=always

.:
1_fastqc
Lab_Notebook.ipynb
raw_data
VarScan.v2.3.9.jar

./1_fastqc:
roommate.fastq
roommate_fastqc.html
roommate_fastqc.zip

./raw_data:
roommate.fastq
roommate.fastq.gz


$~$

## 2. Align your roommate’s data to the reference sequence

Download reference gene manually from [Influenza A virus segment 4 hemagglutinin (HA) gene](https://www.ncbi.nlm.nih.gov/nuccore/KF848938.1?report=fasta) (raw_data/ref_seq.fasta)  

In [52]:
# work in 2_indexed_ref_seq/
!mkdir 2.1_indexed_ref_seq
!cp raw_data/ref_seq.fasta 2_indexed_ref_seq

In [53]:
# INDEX file with reference gene
!bwa index 2.1_indexed_ref_seq/ref_seq.fasta

[bwa_index] Pack FASTA... 0.00 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 0.00 seconds elapse.
[bwa_index] Update BWT... 0.00 sec
[bwa_index] Pack forward-only FASTA... 0.00 sec
[bwa_index] Construct SA from BWT and Occ... 0.00 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index 2_indexed_ref_seq/ref_seq.fasta
[main] Real time: 0.669 sec; CPU: 0.003 sec


In [54]:
# 1. bwa mem:       get alignment from indexed reference gene (.sam)
# 2. samtools view: compress alignment (.sam > .bam)
# 3. samtools sort: sort compressed alignment (.bam) and move it to 2.2_align/

!mkdir 2.2_align
!bwa mem 2.1_indexed_ref_seq/ref_seq.fasta roommate.fastq | samtools view -S -b - | samtools sort - -o 2.2_align/roommate.bam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 68388 sequences (10000129 bp)...
[M::process] read 67628 sequences (10000233 bp)...
[M::mem_process_seqs] Processed 68388 reads in 2.298 CPU sec, 2.233 real sec
[M::process] read 67698 sequences (10000046 bp)...
[M::mem_process_seqs] Processed 67628 reads in 1.750 CPU sec, 1.634 real sec
[M::process] read 67652 sequences (10000169 bp)...
[M::mem_process_seqs] Processed 67698 reads in 2.309 CPU sec, 2.199 real sec
[M::process] read 68072 sequences (10000295 bp)...
[M::mem_process_seqs] Processed 67652 reads in 2.406 CPU sec, 2.307 real sec
[M::process] read 18827 sequences (2716992 bp)...
[M::mem_process_seqs] Processed 68072 reads in 1.980 CPU sec, 1.943 real sec
[M::mem_process_seqs] Processed 18827 reads in 0.963 CPU sec, 0.946 real sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa mem 2.1_indexed_ref_seq/ref_seq.fasta roommate.fastq
[main] Real time: 11.483 sec; CPU: 11.794 sec


In [55]:
# alignment statistics
!samtools flagstat 2.2_align/roommate.bam

361349 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
3084 + 0 supplementary
0 + 0 duplicates
361116 + 0 mapped (99.94% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)


$~$

Alignment went pretty good.

$~$

In [74]:
# Index .bam файл (.bam.bai created)
!samtools index 2.2_align/roommate.bam

In [75]:
# file structure
!ls -R --color=always -X -x

.:
1_fastqc       raw_data        2.1_indexed_ref_seq  2.2_align
ref_seq.fasta  roommate.fastq  Lab_Notebook.ipynb   VarScan.v2.3.9.jar

./1_fastqc:
roommate.fastq  roommate_fastqc.html  roommate_fastqc.zip

./raw_data:
ref_seq.fasta  roommate.fastq  roommate.fastq.gz

./2.1_indexed_ref_seq:
ref_seq.fasta.amb  ref_seq.fasta.ann  ref_seq.fasta.bwt  ref_seq.fasta
ref_seq.fasta.pac  ref_seq.fasta.sa

./2.2_align:
roommate.bam.bai  roommate.bam


$~$

## 3. Look for common variants with VarScan

Look for high frequency variants... (if 95% reads have a mutation)

In [76]:
!mkdir 3_varscan_0.95

$~$

1665 b.p. in reference gene <br>
According to samtools flagstat, there are 361116 mapped reads. <br>
Read length ~35-151 (from fastqc).

So, coverage is about $\frac{361116*151}{1665} \approx 32750$

We will use that value as **--max-depth** parameter in **samtools mpileup**.

In [78]:
# 1. samtools mpileup: Create intermediate (.mpileup) file with info about how many reads
#                      have a mutation at the same position
#                      -d -- max per-file depth (8000 by default)
#
# 2. mpileup2snp:      Review SNPs 
#                      --min-var-freq -- the minimum % of non-reference bases at a position 
#                                        required to call it a mutation in the sample
#                      --variants     -- only output positions that are above our threshold
#                      --output-vcf 1 -- output in .vcf format

!samtools mpileup -d 40000 -f ref_seq.fasta 2.2_align/roommate.bam | java -jar VarScan.v2.3.9.jar mpileup2snp --min-var-freq 0.95 --variants --output-vcf 1 > 3_varscan_0.95/VarScan_results_95.vcf

Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.95
Min avg qual:	15
P-value thresh:	0.01
[mpileup] 1 samples in 1 input files
Input stream not ready, waiting for 5 seconds...
Reading input from STDIN
1665 bases in pileup file
5 variant positions (5 SNP, 0 indel)
0 were failed by the strand-filter
5 variant positions reported (5 SNP, 0 indel)


In [85]:
# file structure
!ls -R --color=always -X -x

.:
1_fastqc            raw_data            2.1_indexed_ref_seq  2.2_align
3_varscan_0.95      ref_seq.fasta.fai   ref_seq.fasta        roommate.fastq
Lab_Notebook.ipynb  VarScan.v2.3.9.jar

./1_fastqc:
roommate.fastq  roommate_fastqc.html  roommate_fastqc.zip

./raw_data:
ref_seq.fasta  roommate.fastq  roommate.fastq.gz

./2.1_indexed_ref_seq:
ref_seq.fasta.amb  ref_seq.fasta.ann  ref_seq.fasta.bwt  ref_seq.fasta
ref_seq.fasta.pac  ref_seq.fasta.sa

./2.2_align:
roommate.bam.bai  roommate.bam

./3_varscan_0.95:
VarScan_results_95.vcf


In [86]:
# awk: parsing tool
# NR>24  -- print only if the line number is >24 (first 24 lines are not informative)
# $1,... -- columns to print

!cat 3_varscan_0.95/VarScan_results_95.vcf | awk 'NR>24 {print $1, $2, $4, $5}'

KF848938.1 72 A G
KF848938.1 117 C T
KF848938.1 774 T C
KF848938.1 999 C T
KF848938.1 1260 A C


$~$

Looking at coding frame and amino acids using [ncbi browser](https://www.ncbi.nlm.nih.gov/nuccore/KF848938.1?report=graph)...

|  position  |    base change    |       aa change       |
|------------|-------------------|-----------------------|
|      72    | A $\rightarrow$ G | Thr $\rightarrow$ Thr |
|     117    | C $\rightarrow$ T | Ala $\rightarrow$ Ala |
|     774    | T $\rightarrow$ C | Phe $\rightarrow$ Phe |
|     999    | C $\rightarrow$ T | Gly $\rightarrow$ Gly |
|    1260    | A $\rightarrow$ C | Leu $\rightarrow$ Leu |

All substitutions are synonymous (silent).<br>
This mutations could not let roommate’s virus to escape the antibodies in my body from the flu vaccine ((9(

$~$

## 4. Look for rare variants with VarScan

Now we gonna look for all rare variants (if 0.1% reads have a mutation).

In [87]:
!mkdir 4_varscan_0.001

In [91]:
# -d 0  -- set max depth to maximum possible value
!samtools mpileup -d 0 -f ref_seq.fasta 2.2_align/roommate.bam | java -jar VarScan.v2.3.9.jar mpileup2snp --min-var-freq 0.001 --variants --output-vcf 1 > 4_varscan_0.001/VarScan_results_01.vcf

[mpileup] 1 samples in 1 input files
[mpileup] Max depth set to maximum value (2147483647)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Input stream not ready, waiting for 5 seconds...
Reading input from STDIN
1665 bases in pileup file
23 variant positions (21 SNP, 2 indel)
0 were failed by the strand-filter
21 variant positions reported (21 SNP, 0 indel)


$~$

We can look at FREQ (variant allele frequency) in the last column of *VarScan_results_01.vcf* file.<br>
They are all about $0.17 - 1\%$ (except described in 3.).

In [106]:
!cat 4_varscan_0.001/VarScan_results_01.vcf | awk 'NR>24 {split($10,freq,":"); print $2, $4, $5, freq[7]}'

72 A G 99.96%
117 C T 99.82%
254 A G 0.17%
276 A G 0.17%
307 C T 0.94%
340 T C 0.17%
389 T C 0.22%
691 A G 0.17%
722 A G 0.2%
744 A G 0.17%
774 T C 99.96%
802 A G 0.23%
859 A G 0.18%
915 T C 0.19%
999 C T 99.86%
1043 A G 0.18%
1086 A G 0.21%
1213 A G 0.22%
1260 A C 99.94%
1280 T C 0.18%
1458 T C 0.84%


In [143]:
# file structure
!ls -R --color=always

.:
1_fastqc             4_varscan_0.001      ref_seq.fasta
2.1_indexed_ref_seq  Lab_Notebook2.ipynb  ref_seq.fasta.fai
2.2_align            Lab_Notebook.ipynb   roommate.fastq
3_varscan_0.95       raw_data             VarScan.v2.3.9.jar

./1_fastqc:
roommate.fastq  roommate_fastqc.html  roommate_fastqc.zip

./2.1_indexed_ref_seq:
ref_seq.fasta      ref_seq.fasta.ann  ref_seq.fasta.pac
ref_seq.fasta.amb  ref_seq.fasta.bwt  ref_seq.fasta.sa

./2.2_align:
roommate.bam  roommate.bam.bai

./3_varscan_0.95:
VarScan_results_95.vcf

./4_varscan_0.001:
VarScan_results_01.vcf

./raw_data:
ref_seq.fasta  roommate.fastq  roommate.fastq.gz


$~$

## 5. Inspect and align the control sample sequencing data

We will look through three controls (from sequencing of isogenic reference samples) -- the result of 3 sequences of one 100% pure sample of the standard (reference) H3N2 influenza virus, PCR amplified, and subcloned into a plasmid. Any “mutations” in the control samples which don’t contain any true genetic variants must be due to errors. 

1. Get all files

In [151]:
# Download files
!wget -c ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/008/SRR1705858/SRR1705858.fastq.gz
!wget -c ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/009/SRR1705859/SRR1705859.fastq.gz
!wget -c ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/000/SRR1705860/SRR1705860.fastq.gz

# Move them to raw_data/ and rename
!mv SRR1705858.fastq.gz raw_data/control1.fastq.gz
!mv SRR1705859.fastq.gz raw_data/control2.fastq.gz
!mv SRR1705860.fastq.gz raw_data/control3.fastq.gz

# Unpack
!gunzip -k raw_data/control*

--2021-10-24 20:29:03--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/008/SRR1705858/SRR1705858.fastq.gz
Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138
Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 14782243 (14M) [application/octet-stream]
Saving to: ‘SRR1705858.fastq.gz’

SRR1705858.fastq.gz 100%[===================>]  14,10M   617KB/s    in 26s     

2021-10-24 20:29:34 (555 KB/s) - ‘SRR1705858.fastq.gz’ saved [14782243/14782243]

--2021-10-24 20:29:34--  http://ftp.sra.ebi.ac.uk/vol1/fastq/SRR170/009/SRR1705859/SRR1705859.fastq.gz
Resolving ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)... 193.62.193.138
Connecting to ftp.sra.ebi.ac.uk (ftp.sra.ebi.ac.uk)|193.62.193.138|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 13581641 (13M) [application/octet-stream]
Saving to: ‘SRR1705859.fastq.gz’

SRR1705859.fastq.gz 100%[===================>]  12,95M   666KB/s   

2. Look how many reads are in each control file:

In [226]:
!cat raw_data/control1.fastq | wc -l | awk '{print $1/4 " reads in file 1"}'
!cat raw_data/control2.fastq | wc -l | awk '{print $1/4 " reads in file 2"}'
!cat raw_data/control3.fastq | wc -l | awk '{print $1/4 " reads in file 3"}'

256586 reads in file 1
233327 reads in file 2
249964 reads in file 3


3. Mean read lengths in each file:

In [227]:
# print legth of each 4th row, beginning from 2nd --> 
# we want to calculate mean length = (length of all reads)/(number of reads)

!cat raw_data/control1.fastq | awk '{if((NR+2)%4==0) sum+=length;}END{print sum/256586;}'
!cat raw_data/control2.fastq | awk '{if((NR+2)%4==0) sum+=length;}END{print sum/233327;}' 
!cat raw_data/control3.fastq | awk '{if((NR+2)%4==0) sum+=length;}END{print sum/249964;}' 

148,561
148,446
148,703


So again, there are 1665 b.p. in reference gene. <br>
There are 256586, 233327 and 249964 reads in control files 1,2 and 3. <br>
Mean read length is about 148.5.

All that info gives us rough estimate of the coverage in our samples $\frac{256586*148.5}{1665} \approx 22885$ (for file control1.fastq).

In [208]:
!mkdir 5_align_control

In [228]:
# 1. bwa mem:       make alignment based on ref seq KF848938.1 (.sam)
# 2. samtools view: compress alignment file (.sam > .bam)
# 3. samtools sort: sort compress file (.bam) and save it to 5_align_control/

!bwa mem 2.1_indexed_ref_seq/ref_seq.fasta raw_data/control1.fastq | samtools view -S -b - | samtools sort - -o 5_align_control/control_alignment1.bam
!bwa mem 2.1_indexed_ref_seq/ref_seq.fasta raw_data/control2.fastq | samtools view -S -b - | samtools sort - -o 5_align_control/control_alignment2.bam
!bwa mem 2.1_indexed_ref_seq/ref_seq.fasta raw_data/control3.fastq | samtools view -S -b - | samtools sort - -o 5_align_control/control_alignment3.bam

[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 67448 sequences (10000210 bp)...
[M::process] read 67230 sequences (10000094 bp)...
[M::mem_process_seqs] Processed 67448 reads in 5.779 CPU sec, 5.549 real sec
[M::process] read 67100 sequences (10000113 bp)...
[M::mem_process_seqs] Processed 67230 reads in 6.470 CPU sec, 6.195 real sec
[M::process] read 54808 sequences (8118313 bp)...
[M::mem_process_seqs] Processed 67100 reads in 6.567 CPU sec, 6.382 real sec
[M::mem_process_seqs] Processed 54808 reads in 5.547 CPU sec, 5.447 real sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa mem 2.1_indexed_ref_seq/ref_seq.fasta raw_data/control1.fastq
[main] Real time: 25.112 sec; CPU: 24.751 sec
[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 67476 sequences (10000274 bp)...
[M::process] read 67236 sequences (10000039 bp)...
[M::mem_process_seqs] Processed 67476 reads in 5.951 CPU sec, 5.763 real sec
[M::process] read 67216 sequences (10000283 bp)...
[M::mem_proc

In [229]:
# Index each .bam file with control alignments (creates .bam.bai)
!samtools index 5_align_control/control_alignment1.bam
!samtools index 5_align_control/control_alignment2.bam
!samtools index 5_align_control/control_alignment3.bam

In [211]:
# file structure
!ls -R --color=always

.:
1_fastqc             5_align_control      ref_seq.fasta
2.1_indexed_ref_seq  labjournal.ipynb     ref_seq.fasta.fai
2.2_align            Lab_Notebook2.ipynb  roommate.fastq
3_varscan_0.95       Lab_Notebook.ipynb   VarScan.v2.3.9.jar
4_varscan_0.001      raw_data

./1_fastqc:
roommate.fastq  roommate_fastqc.html  roommate_fastqc.zip

./2.1_indexed_ref_seq:
ref_seq.fasta      ref_seq.fasta.ann  ref_seq.fasta.pac
ref_seq.fasta.amb  ref_seq.fasta.bwt  ref_seq.fasta.sa

./2.2_align:
roommate.bam  roommate.bam.bai

./3_varscan_0.95:
VarScan_results_95.vcf

./4_varscan_0.001:
VarScan_results_01.vcf

./5_align_control:
control_alignment1.bam      control_alignment2.bam.bai
control_alignment1.bam.bai  control_alignment3.bam
control_alignment2.bam      control_alignment3.bam.bai

./raw_data:
control1.fastq     control2.fastq.gz  ref_seq.fasta
control1.fastq.gz  control3.fastq     roommate.fastq
control2.fastq     control3.fastq.gz  roommate.fastq.gz


$~$

## 6. Use VarScan to look for rare variants in the reference files.

In [215]:
!mkdir 6_varscan_control

Now we look for "mutations" in control files (we know they are actually method errors).

In [230]:
!samtools mpileup -d 0 -f ref_seq.fasta 5_align_control/control_alignment1.bam | java -jar VarScan.v2.3.9.jar mpileup2snp --min-var-freq 0.001 --variants --output-vcf 1 > 6_varscan_control/VarScan_control1_01.vcf

[mpileup] 1 samples in 1 input files
[mpileup] Max depth set to maximum value (2147483647)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Input stream not ready, waiting for 5 seconds...
Input stream not ready, waiting for 5 seconds...
Reading input from STDIN
1665 bases in pileup file
58 variant positions (58 SNP, 0 indel)
1 were failed by the strand-filter
57 variant positions reported (57 SNP, 0 indel)


In [231]:
!samtools mpileup -d 0 -f ref_seq.fasta 5_align_control/control_alignment2.bam | java -jar VarScan.v2.3.9.jar mpileup2snp --min-var-freq 0.001 --variants --output-vcf 1 > 6_varscan_control/VarScan_control2_01.vcf

[mpileup] 1 samples in 1 input files
[mpileup] Max depth set to maximum value (2147483647)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Input stream not ready, waiting for 5 seconds...
Input stream not ready, waiting for 5 seconds...
Reading input from STDIN
1665 bases in pileup file
54 variant positions (54 SNP, 0 indel)
2 were failed by the strand-filter
52 variant positions reported (52 SNP, 0 indel)


In [232]:
!samtools mpileup -d 0 -f ref_seq.fasta 5_align_control/control_alignment3.bam | java -jar VarScan.v2.3.9.jar mpileup2snp --min-var-freq 0.001 --variants --output-vcf 1 > 6_varscan_control/VarScan_control3_01.vcf

[mpileup] 1 samples in 1 input files
[mpileup] Max depth set to maximum value (2147483647)
Only SNPs will be reported
Min coverage:	8
Min reads2:	2
Min var freq:	0.001
Min avg qual:	15
P-value thresh:	0.01
Input stream not ready, waiting for 5 seconds...
Input stream not ready, waiting for 5 seconds...
Reading input from STDIN
1665 bases in pileup file
61 variant positions (61 SNP, 0 indel)
0 were failed by the strand-filter
61 variant positions reported (61 SNP, 0 indel)


Export all variants from control files and from roommates sequence results:

In [234]:
# control files
!cat 6_varscan_control/VarScan_control1_01.vcf | awk 'NR>24 {split($10,freq,":"); print $2, $4, $5, freq[7]}' > 6_varscan_control/stat_1.txt
!cat 6_varscan_control/VarScan_control2_01.vcf | awk 'NR>24 {split($10,freq,":"); print $2, $4, $5, freq[7]}' > 6_varscan_control/stat_2.txt
!cat 6_varscan_control/VarScan_control3_01.vcf | awk 'NR>24 {split($10,freq,":"); print $2, $4, $5, freq[7]}' > 6_varscan_control/stat_3.txt

In [235]:
# roommtate's file
!cat 4_varscan_0.001/VarScan_results_01.vcf | awk 'NR>24 {split($10,freq,":"); print $2, $4, $5, freq[7]}' > 6_varscan_control/roommate_stat.txt

In [70]:
# file structure
# !ls -R --color=always

$~$

## 7. Compare the control results to your roommate’s results

First, read our stat files...

In [40]:
import pandas as pd

In [54]:
# roommate stats

roommate_mutations = pd.read_csv('6_varscan_control/roommate_stat.txt', 
                                 sep=' ', 
                                 names = ['pos', 'ref', 'alt', 'freq'])
roommate_mutations.head()

pos ref alt    freq
0   72   A   G  99.96%
1  117   C   T  99.82%
2  254   A   G   0.17%
3  276   A   G   0.17%
4  307   C   T   0.94%

In [55]:
# stats from control files

control_files = ['stat_1.txt', 'stat_2.txt', 'stat_3.txt']
controls = {
    file:pd.read_csv('6_varscan_control/'+file, 
                     sep=' ', 
                     names = ['pos', 'ref', 'alt', 'freq']) 
    for file in control_files
}

In [56]:
# merge controls 1 and 2
controls_12 = pd.merge(controls['stat_1.txt'], controls['stat_2.txt'], 
                       on=['pos','ref','alt'],
                       how='outer', 
                       suffixes=('_1', '_2'))

# merge 1,2 and control№3
controls_123 = pd.merge(controls_12, controls['stat_3.txt'], 
                        on=['pos','ref','alt'],
                        how='outer')

# rename column
controls_123.columns.values[5] = 'freq_3'

controls_123

pos ref alt freq_1 freq_2 freq_3
0     38   T   C  0.66%    NaN   0.7%
1     54   T   C   0.3%    NaN    NaN
2     72   A   G   0.3%    NaN    NaN
3     95   A   G  0.24%    NaN  0.24%
4    117   C   T   0.3%    NaN    NaN
..   ...  ..  ..    ...    ...    ...
80   271   A   G    NaN    NaN  0.21%
81   759   T   C    NaN    NaN  0.19%
82  1105   A   G    NaN    NaN  0.22%
83  1209   A   G    NaN    NaN  0.27%
84  1301   A   G    NaN    NaN  0.22%

[85 rows x 6 columns]

In [57]:
# merge stats from control files with roommate info
control_mate = pd.merge(controls_123, roommate_mutations, 
                        on=['pos','ref','alt'],
                        how='outer')

# delete all % symbols
control_mate = control_mate.replace(['\%'], [''], regex=True)

# ddof = 0 cause otherwise we will get zero division and NaN in std_freq row when it is only one freq
# control_mate['mean_freq'] = control_mate[['freq_1','freq_2','freq_3_x']].astype(float).mean(axis=1, skipna = True)
# control_mate['std_freq'] = control_mate[['freq_1','freq_2','freq_3_x']].astype(float).std(axis=1, skipna = True, ddof=0)
control_mate.head(10)

pos ref alt freq_1 freq_2 freq_3_x   freq
0   38   T   C   0.66    NaN      0.7    NaN
1   54   T   C    0.3    NaN      NaN    NaN
2   72   A   G    0.3    NaN      NaN  99.96
3   95   A   G   0.24    NaN     0.24    NaN
4  117   C   T    0.3    NaN      NaN  99.82
5  165   T   C   0.24   0.27     0.25    NaN
6  183   A   G    0.3   0.22     0.23    NaN
7  216   A   G   0.22   0.24     0.24    NaN
8  218   A   G   0.28   0.29     0.23    NaN
9  222   T   C   0.26   0.25      0.3    NaN

The **average** and **standard deviation** of the frequencies reported in control files are:

In [69]:
means = control_mate[['freq_1','freq_2','freq_3_x']].astype(float).mean(axis=0, skipna = True)
stds = control_mate[['freq_1','freq_2','freq_3_x']].astype(float).std(axis=0, skipna = True)

for (m, s, i) in zip(means, stds, [1, 2, 3]):
    print(f'For control file {i}: freq = {m:.2f}±{s:.2f}')

For control file 1: freq = 0.26±0.07
For control file 2: freq = 0.24±0.05
For control file 3: freq = 0.25±0.08


In [66]:
!mkdir 7_roommate_vs_control

In [68]:
control_mate.to_csv('7_roommate_vs_control/statistics.csv')

Average frequency from all control files is 0.25±0.07.<br>
Now we can visually look at column **freq** (frequency of each variant in sequence results of our roommate) in file *statistics.csv* and compare it with average control frequency. <br>
We found 2 mutations (*see table below*), that are present in roommate's sequence data, but not in control data. This rare mutation's frequencies are more than 3 standard deviations away from the averages in the reference (control) files (>0.25+0.21). <br>
So, we assume, that this variants appear not because of PCR/sequencing errors, but because of virus mutations in roommate's cells.

|  position  |    base change    | freq |       aa change       |
|------------|-------------------|------|-----------------------|
|     307    | C $\rightarrow$ T | 0.94 |       Pro103Ser       |
|    1458    | T $\rightarrow$ C | 0.84 |       Tyr486Tyr       |

Only one of this 2 mutations leads to amino acid change (from non-polar Proline to polar Serine).

$~$

## 8. Epitope mapping

An **epitope** is the part of an antigen that is recognized by the immune system.<br>
Using [Munoz et al, 2004](https://drive.google.com/file/d/1xe5-4LxIV4bO4mX6jhvrMAqtpOpkWsXm/view) we can determine if Pro103Ser mutation from roommate’s flu infection is located in an epitope region of hemagglutinin.

**RESULT:** One missense mutation in epitope D

## Some extra calculations (for myself)

In [1]:
!samtools flagstat 5_align_control/control_alignment1.bam
!samtools flagstat 5_align_control/control_alignment2.bam
!samtools flagstat 5_align_control/control_alignment3.bam

256744 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
158 + 0 supplementary
0 + 0 duplicates
256658 + 0 mapped (99.97% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
233451 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
124 + 0 supplementary
0 + 0 duplicates
233375 + 0 mapped (99.97% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)
0 + 0 with itself and mate mapped
0 + 0 singletons (N/A : N/A)
0 + 0 with mate mapped to a different chr
0 + 0 with mate mapped to a different chr (mapQ>=5)
250184 + 0 in total (QC-passed reads + QC-failed reads)
0 + 0 secondary
220 + 0 supplementary
0 + 0 duplicates
250108 + 0 mapped (99.97% : N/A)
0 + 0 paired in sequencing
0 + 0 read1
0 + 0 read2
0 + 0 properly paired (N/A : N/A)

In [4]:
!samtools view -f4 5_align_control/control_alignment3.bam

SRR1705860.98	4	*	0	0	*	*	0	0	GTTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGTTA	????ABBBDDDEDDDDGGGGGGIHGHHHHHHHFHIHHFEGHHIIHHHHHHIHHGGFFFFFHHIIIGHHHGGHIHGHI	AS:i:0	XS:i:0
SRR1705860.332	4	*	0	0	*	*	0	0	ACCATGAAGACTATCATTGCTTTGAGCTACATTCCATGTCTGGTTTTCGCTCAAAAACTTCCTGGA	??AAABBBDDDDDDDEGGGGGGIIIIIIIIIIIIIIIIGHIHHCGHIIHHHIHIIIIHIIIIIIII	AS:i:0	XS:i:0
SRR1705860.497	4	*	0	0	*	*	0	0	GTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGT	?????BBBDDDDDDDDGGGGGGIIHHHHHIHFHIIIIGHFHIHCFHGHHHHHHIHHHHHHIFEGFHHHGGBG	AS:i:0	XS:i:0
SRR1705860.498	4	*	0	0	*	*	0	0	CATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGT	???A?BBBDBDDDDDDGGGGGGHHHHIIEHHIIHHHIG	AS:i:0	XS:i:0
SRR1705860.519	4	*	0	0	*	*	0	0	GTGCTATTGTCATTTCCAGGAAGTTTTTGAGCGAAAACCAGACATAGAATGTAGCTCAAAGCAATGATAGTCTTCATGGT	????ABBBEEEDDDDDGGGGGGIFHIIIHHIIHHHHHHHIIIIIIIIIIIIHIHIIIIIIHHHIIIIIIIFHIIIIHHIF	AS:i:0	XS:i:0
SRR1705860.881	4	*	0	0	*	*	0	0	CCATGAAGACTATCATTGCTTTGAGCTACATTCTATGTCTGGTTTTCGCTCAAAAACTTCCTGGAAATG